File: java21/src/main/java/io/github/sps23/designpatterns/command/DocumentEditor.java
package io.github.sps23.designpatterns.command;

import java.util.ArrayDeque;
import java.util.Deque;

public final class DocumentEditor {
    private final StringBuilder content = new StringBuilder();

    public void insert(int index, String value) {
        if (index < 0 || index > content.length()) {
            throw new IndexOutOfBoundsException("Index out of range: " + index);
        }
        content.insert(index, value);
    }

    public void delete(int start, int end) {
        if (start < 0 || end < start || end > content.length()) {
            throw new IndexOutOfBoundsException("Range is invalid: [" + start + ", " + end + "]");
        }
        content.delete(start, end);
    }

    public String text() {
        return content.toString();
    }
}

interface Command {
    void execute();

    void undo();
}

final class InsertTextCommand implements Command {
    private final DocumentEditor editor;
    private final int index;
    private final String value;

    InsertTextCommand(DocumentEditor editor, int index, String value) {
        this.editor = editor;
        this.index = index;
        this.value = value;
    }

    @Override
    public void execute() {
        editor.insert(index, value);
    }

    @Override
    public void undo() {
        editor.delete(index, index + value.length());
    }
}

final class CommandHistory {
    private final Deque<Command> undoStack = new ArrayDeque<>();
    private final Deque<Command> redoStack = new ArrayDeque<>();

    public void execute(Command command) {
        command.execute();
        undoStack.push(command);
        redoStack.clear();
    }

    public void undo() {
        if (undoStack.isEmpty()) {
            return;
        }
        var command = undoStack.pop();
        command.undo();
        redoStack.push(command);
    }

    public void redo() {
        if (redoStack.isEmpty()) {
            return;
        }
        var command = redoStack.pop();
        command.execute();
        undoStack.push(command);
    }

    public boolean canUndo() {
        return !undoStack.isEmpty();
    }

    public boolean canRedo() {
        return !redoStack.isEmpty();
    }
}

File: kotlin/src/main/kotlin/io/github/sps23/designpatterns/command/DocumentEditor.kt
package io.github.sps23.designpatterns.command

import java.util.ArrayDeque
import java.util.Deque

class DocumentEditor {
    private val content = StringBuilder()

    fun insert(index: Int, value: String) {
        require(index in 0..content.length) { "Index out of range: $index" }
        content.insert(index, value)
    }

    fun delete(start: Int, end: Int) {
        require(start in 0..end && end <= content.length) { "Range is invalid: [$start, $end]" }
        content.delete(start, end)
    }

    fun text(): String = content.toString()
}

interface Command {
    fun execute()
    fun undo()
}

class InsertTextCommand(
    private val editor: DocumentEditor,
    private val index: Int,
    private val value: String,
) : Command {
    override fun execute() {
        editor.insert(index, value)
    }

    override fun undo() {
        editor.delete(index, index + value.length)
    }
}

class CommandHistory {
    private val undoStack: Deque<Command> = ArrayDeque()
    private val redoStack: Deque<Command> = ArrayDeque()

    fun execute(command: Command) {
        command.execute()
        undoStack.push(command)
        redoStack.clear()
    }

    fun undo() {
        if (undoStack.isEmpty()) return
        val command = undoStack.pop()
        command.undo()
        redoStack.push(command)
    }

    fun redo() {
        if (redoStack.isEmpty()) return
        val command = redoStack.pop()
        command.execute()
        undoStack.push(command)
    }

    fun canUndo(): Boolean = undoStack.isNotEmpty()
    fun canRedo(): Boolean = redoStack.isNotEmpty()
}
